README.rst
.. initialize RKM codes

    >>> from rkm_codes import set_prefs
    >>> set_prefs(
    ...     rkm_maps=None, units_to_rkm_base_code=None, map_sf=None,
    ...     show_units=None, strip_zeros=None, minus_sign=None, prec=None
    ... )

RKM codes
=========

| Version: 0.4.0
| Released: 2019-08-29
|

.. image:: https://img.shields.io/travis/KenKundert/rkm_codes/master.svg
    :target: https://travis-ci.org/KenKundert/rkm_codes

.. image:: https://img.shields.io/coveralls/KenKundert/rkm_codes.svg
    :target: https://coveralls.io/r/KenKundert/rkm_codes

.. image:: https://img.shields.io/pypi/v/rkm_codes.svg
    :target: https://pypi.python.org/pypi/rkm_codes

.. image:: https://img.shields.io/pypi/pyversions/rkm_codes.svg
    :target: https://pypi.python.org/pypi/rkm_codes/

RKM codes are used to represent electrical quantities in labels, particularly on
schematics and on the components themselves.  They are standardized in various
national and international standards, including: IEC 60062 (1952) (formerly IEC 62),
DIN 40825 (1973), BS 1852 (1974), IS 8186 (1976) and EN 60062 (1993).
IEC-60062 was significantly updated in 2016.

RKM codes were originally used as part marking codes.  This shorthand
notation is widely used in electrical engineering to denote the values of
resistors and capacitors in circuit diagrams and in the production of electronic
circuits (for example in bills of material and in silk screens). This method
avoids overlooking the decimal separator, which may not be rendered reliably on
components or when duplicating documents.  They also provide the benefit that
the characters within a RKM code are either letters or digits, and so can be
embedded within identifiers without introducing invalid characters.

IEC 60062 is described in https://en.wikipedia.org/wiki/RKM_code.

Essentially an RKM version of a number is the number with a scale factor where
the decimal point replaced by the scale factor. For example, a resistance of
4.7kΩ becomes 4k7. If there is no scale factor, the decimal point is replaced by
a letter that signifies the type of the component.  For example, a resistance of
4.7Ω becomes 4r7.

Resistance examples:

    | R47 → 0.47 Ω
    | 4R7 → 4.7 Ω
    | 470R → 470 Ω
    | 4K7 → 4.7 kΩ
    | 47K → 47 kΩ
    | 47K3 → 47.3 kΩ
    | 470K → 470 kΩ
    | 4M7 → 4.7 MΩ

In the standard, large values are assumed to be resistances and small values are
assumed to be capacitances.  So 4k7 is a resistance and 2n5 is a capacitance.
However, this package also supports a version of RKM codes where the units are
not implied by the value, making RKM codes suitable for a wider variety of value
types, such as voltage, current, and inductance.


Installing
----------

This package converts RKM codes to `QuantiPhy Quantities 
<https://quantiphy.readthedocs.io>`_ and Quantities to RKM codes.

Install with::

    pip3 install --user rkm_codes

Requires Python3.4 or better.


Converting to and from RKM Codes
--------------------------------

The following is a simple example of how to convert back and forth between RKM 
codes and Quantities::

    >>> from rkm_codes import from_rkm, to_rkm
    >>> r = from_rkm('6K8')
    >>> r
    Quantity('6.8k')

    >>> to_rkm(r)
    '6K8'

Notice that in this case the quantity does not include units. That is because by 
default *rkm_codes* assumes unitless numbers. You can change this behavior.  Out 
of the box *rkm_codes* supports two kinds of numbers, unitless and those that 
follow the IEC60062 standard. You can switch between those two kinds of numbers 
using something like this::

    >>> from rkm_codes import set_prefs, IEC60062_MAPS, UNITLESS_MAPS
    >>> r = from_rkm('6k8')
    >>> r
    Quantity('6.8k')

    >>> set_prefs(rkm_maps=IEC60062_MAPS)
    >>> from_rkm('6k8')
    Quantity('6.8 kΩ')

    >>> set_prefs(rkm_maps=UNITLESS_MAPS)
    >>> from_rkm('6k8')
    Quantity('6.8k')

In either case, *rkm_codes* allows you to explicitly specify the units, which 
always overrides any implied units::

    >>> set_prefs(rkm_maps=UNITLESS_MAPS)
    >>> from_rkm('6kΩ8')
    Quantity('6.8 kΩ')

    >>> i = from_rkm('2uA5')
    >>> i
    Quantity('2.5 uA')

When converting to an RKM code, you can instruct that the units be included::

    >>> to_rkm(i, show_units=True)
    '2μA5'

You can also indicate how many digits should be included::

    >>> to_rkm(i.add(1e-9), prec=5, show_units=True)
    '2μA501'

Normally, any excess zeros are removed, but you can change that too::

    >>> to_rkm(i.add(1e-9), prec=5, show_units=True, strip_zeros=False)
    '2μA50100'

You can create your own maps by passing in a dictionary that maps a RKM base 
code character into a scale factor and units. For example, you could create 
a map that uses 'd' or 'D' to represent the decimal point in numbers without 
scale factors rather than 'r', 'c', etc.  For example::

    >>> set_prefs(rkm_maps=dict(d=('', ''), D=('', '')))
    >>> from_rkm('6d8')
    Quantity('6.8')

    >>> from_rkm('2d5')
    Quantity('2.5')

Passing *None* for the value of a map returns it to its default value.

If *rkm_codes* encounters a RKM base code character that is not in the map, it 
simply uses that character. In this way, scale factors are handled::

    >>> from_rkm('6k8')
    Quantity('6.8k')

When converting from Quantities to RKM codes, you can override the default 
mappings from units to RKM base code characters. The default mapping maps 'Ω' 
and 'Ohm' to 'r', 'F' to 'c', 'H' to 'l', 'V' to 'v', and 'A' to 'i'.  However, 
you may prefer uppercase base characters, which is more in alignment with the 
original standard. To get that, you can use something like this::

    >>> rkm_base_code_mappings = {
    ...     'Ω': 'R',
    ...     'Ohm': 'R',
    ...     'F': 'C',
    ...     'H': 'L',
    ...     'V': 'V',
    ...     'A': 'I',
    ... }
    >>> set_prefs(rkm_maps=IEC60062_MAPS, units_to_rkm_base_code=rkm_base_code_mappings)
    >>> r = from_rkm('k0012')
    >>> to_rkm(r)
    '1R2'

You can control the scale factors used by to_rkm() by setting *map_sf* using 
*set_prefs*. The default maps 'u' to 'μ' and 'k' to 'K'. You might wish to 
prevent the use of 'μ' while retaining the use of 'K', which you can do with:

    >>> set_prefs(map_sf=dict(u='µ'))
    >>> c = from_rkm('5u')
    >>> to_rkm(c)
    '5µ'


Finding RKM Codes
-----------------

*find_rkm* is available for finding the RKM codes embedded in text strings.  
Using it, you can iterate through all the numbers specified using RKM:

    >>> from rkm_codes import find_rkm

    >>> text = '''
    ...     An RKM code that may include explicitly specified. Examples of
    ...     acceptable RKM codes for resistance include:   R47 (0.47 Ω), 4R7
    ...     (4.7 Ω), 470R (470 Ω), 4K7 (4.7 kΩ), 47K (47 kΩ), 47K3 (47.3 kΩ),
    ...     470K (470 kΩ), and 4M7 (4.7 MΩ).
    ... '''
    >>> for num in find_rkm(text):
    ...     print(num)
    470 mΩ
    4.7 Ω
    470 Ω
    4.7 kΩ
    47 kΩ
    47.3 kΩ
    470 kΩ
    4.7 MΩ

When the RKM code is not isolated by punctuation or spaces it can get confused 
by leading and trailing text.  You can often resolve this issue by restricting 
the matches to either the leading or trailing digit forms of the RKM code. Do so 
by specifying either 'ld' or 'td' as a second argument.  For example:

    >>> for num in find_rkm('sink200nA'):
    ...     print(num)
    200 msink

    >>> for num in find_rkm('sink200nA', 'ld'):
    ...     print(num)
    200 nA


Pin Name Generator Example
--------------------------

As a practical example of the use of RKM codes, imagine wanting a program that 
creates pin names for an electrical circuit based on a naming convention.  It 
would take a table of pin characteristics that are used to create the names.  
For example::

    >>> from quantiphy import Quantity
    >>> from rkm_codes import to_rkm, set_prefs as set_rkm_prefs

    >>> pins = [
    ...     dict(kind='ibias', direction='out', polarity='sink', dest='dac', value='250nA'),
    ...     dict(kind='ibias', direction='out', polarity='src', dest='rampgen', value='2.5μA'),
    ...     dict(kind='vref', direction='out', dest='dac', value='1.25V'),
    ...     dict(kind='vdda', direction='in', value='2.5V'),
    ... ]
    >>> set_rkm_prefs(map_sf={}, units_to_rkm_base_code=None, show_units=True, prec=2)

    >>> for pin in pins:
    ...     components = []
    ...     if 'value' in pin:
    ...         pin['VALUE'] = to_rkm(Quantity(pin['value']))
    ...     for name in ['dest', 'kind', 'direction', 'VALUE', 'polarity']:
    ...         if name in pin:
    ...             components.append(pin[name])
    ...     print('_'.join(components))
    dac_ibias_out_250nA_sink
    rampgen_ibias_out_2uA5_src
    dac_vref_out_1V25
    vdda_in_2V5


Releases
--------

**Latest development release**:
    | Version: 0.4.0
    | Released: 2019-08-29

**0.4 (2019-08-29)**:
    - added find_rkm()

**0.3 (2019-08-23)**:
    - move the units to the middle of the number with the scale factor
    - added support for signed numbers
    - added *show_units*, *strip_zeros*, *minus_sign*, and *prec* to preferences
    - this release is not backward compatible; units at the end of the number 
      are no longer supported

**0.2 (2018-09-14)**:
    - fixed issue in *set_prefs()*

**0.1 (2018-09-12)**:
    - initial release




releases.rst
Releases
========


**Latest development release**:
    | Version: 0.3.0
    | Released: 2019-08-23

**0.3 (2019-08-23)**:
    - move the units to the middle of the number with the scale factor
    - added support for signed numbers
    - added *show_units*, *strip_zeros*, *minus_sign*, and *prec* to preferences
    - this release is not backward compatible; units at the end of the number 
      are no longer supported

**0.2 (2018-09-14)**:
    - fixed issue in *set_prefs()*

**0.1 (2018-09-12)**:
    - initial release



tests/test_rkm_code.py
# encoding: utf8

from rkm_codes import (
    set_prefs, from_rkm, to_rkm, IEC60062_MAPS, UNITLESS_MAPS
)
from quantiphy import Quantity

testcases = [
    dict(maps=UNITLESS_MAPS, icode='R47', q='470m', ocode='470m', ocodeu='470m'),
    dict(maps=UNITLESS_MAPS, icode='R47Ω', q='470 mΩ', ocode='470m', ocodeu='470mΩ'),
    dict(maps=UNITLESS_MAPS, icode='4R7', q='4.7', ocode='4d7', ocodeu='4d7'),
    dict(maps=UNITLESS_MAPS, icode='470R', q='470', ocode='470', ocodeu='470'),
    dict(maps=UNITLESS_MAPS, icode='4K7', q='4.7k', ocode='4K7', ocodeu='4K7'),
    dict(maps=UNITLESS_MAPS, icode='4K7Ω', q='4.7 kΩ', ocode='4K7', ocodeu='4K7Ω'),
    dict(maps=UNITLESS_MAPS, icode='47K', q='47k', ocode='47K', ocodeu='47K'),
    dict(maps=UNITLESS_MAPS, icode='47K3', q='47.3k', ocode='47K', ocodeu='47K'),
    dict(maps=UNITLESS_MAPS, icode='470K', q='470k', ocode='470K', ocodeu='470K'),
    dict(maps=UNITLESS_MAPS, icode='4M7', q='4.7M', ocode='4M7', ocodeu='4M7'),
    dict(maps=UNITLESS_MAPS, icode='150nA', q='150 nA', ocode='150n', ocodeu='150nA'),
    dict(maps=UNITLESS_MAPS, icode='2v5', q='2.5', ocode='2d5', ocodeu='2d5'),
    dict(maps=UNITLESS_MAPS, icode='2n5A', q='2.5 nA', ocode='2n5', ocodeu='2n5A'),
    dict(maps=UNITLESS_MAPS, icode='f220', q='220m', ocode='220m', ocodeu='220m'),
    dict(maps=UNITLESS_MAPS, icode='2c', q='2', ocode='2', ocodeu='2'),
    dict(maps=UNITLESS_MAPS, icode='2c5', q='2.5', ocode='2d5', ocodeu='2d5'),
    dict(maps=UNITLESS_MAPS, icode='2f', q='2', ocode='2', ocodeu='2'),
    dict(maps=UNITLESS_MAPS, icode='2f5', q='2.5', ocode='2d5', ocodeu='2d5'),
    dict(maps=UNITLESS_MAPS, icode='2p5F', q='2.5 pF', ocode='2p5', ocodeu='2p5F'),
    dict(maps=UNITLESS_MAPS, icode='h220', q='220m', ocode='220m', ocodeu='220m'),
    dict(maps=UNITLESS_MAPS, icode='2l', q='2', ocode='2', ocodeu='2'),
    dict(maps=UNITLESS_MAPS, icode='2l5', q='2.5', ocode='2d5', ocodeu='2d5'),
    dict(maps=UNITLESS_MAPS, icode='2p5H', q='2.5 pH', ocode='2p5', ocodeu='2p5H'),

    dict(maps=IEC60062_MAPS, icode='R47', q='470 mΩ', ocode='470m', ocodeu='470mΩ'),
    dict(maps=IEC60062_MAPS, icode='R47Ω', q='470 mΩ', ocode='470m', ocodeu='470mΩ'),
    dict(maps=IEC60062_MAPS, icode='4R7', q='4.7 Ω', ocode='4r7', ocodeu='4r7Ω'),
    dict(maps=IEC60062_MAPS, icode='470R', q='470 Ω', ocode='470', ocodeu='470Ω'),
    dict(maps=IEC60062_MAPS, icode='4K7', q='4.7 kΩ', ocode='4K7', ocodeu='4K7Ω'),
    dict(maps=IEC60062_MAPS, icode='4K7Ω', q='4.7 kΩ', ocode='4K7', ocodeu='4K7Ω'),
    dict(maps=IEC60062_MAPS, icode='47K', q='47 kΩ', ocode='47K', ocodeu='47KΩ'),
    dict(maps=IEC60062_MAPS, icode='47K3', q='47.3 kΩ', ocode='47K', ocodeu='47KΩ'),
    dict(maps=IEC60062_MAPS, icode='470K', q='470 kΩ', ocode='470K', ocodeu='470KΩ'),
    dict(maps=IEC60062_MAPS, icode='4M7', q='4.7 MΩ', ocode='4M7', ocodeu='4M7Ω'),
    dict(maps=IEC60062_MAPS, icode='150nA', q='150 nA', ocode='150n', ocodeu='150nA'),
    dict(maps=IEC60062_MAPS, icode='2v5', q='2.5 v', ocode='2v5', ocodeu='2v5v'),
    dict(maps=IEC60062_MAPS, icode='2n5A', q='2.5 nA', ocode='2n5', ocodeu='2n5A'),
    dict(maps=IEC60062_MAPS, icode='f220', q='220a', ocode='220a', ocodeu='220a'),
    dict(maps=IEC60062_MAPS, icode='2c', q='2 F', ocode='2', ocodeu='2F'),
    dict(maps=IEC60062_MAPS, icode='2c5', q='2.5 F', ocode='2c5', ocodeu='2c5F'),
    dict(maps=IEC60062_MAPS, icode='2f', q='2f', ocode='2f', ocodeu='2f'),
    dict(maps=IEC60062_MAPS, icode='2f5', q='2.5f', ocode='2f5', ocodeu='2f5'),
    dict(maps=IEC60062_MAPS, icode='2p5F', q='2.5 pF', ocode='2p5', ocodeu='2p5F'),
    dict(maps=IEC60062_MAPS, icode='h220', q='220 mh', ocode='220m', ocodeu='220mh'),
    dict(maps=IEC60062_MAPS, icode='2l', q='2 mΩ', ocode='2m', ocodeu='2mΩ'),
    dict(maps=IEC60062_MAPS, icode='2l5', q='2.5 mΩ', ocode='2m5', ocodeu='2m5Ω'),
    dict(maps=IEC60062_MAPS, icode='2p5H', q='2.5 pH', ocode='2p5', ocodeu='2p5H'),

    dict(maps=IEC60062_MAPS, icode='l047', q='47 uΩ', ocode='47μ', ocodeu='47μΩ'),  # using invalid sf
    dict(maps=IEC60062_MAPS, icode='l47', q='470 uΩ', ocode='470μ', ocodeu='470μΩ'), # using invalid sf
    dict(maps=IEC60062_MAPS, icode='4l7', q='4.7 mΩ', ocode='4m7', ocodeu='4m7Ω'),
    dict(maps=IEC60062_MAPS, icode='47l', q='47 mΩ', ocode='47m', ocodeu='47mΩ'),
    dict(maps=IEC60062_MAPS, icode='470l', q='470 mΩ', ocode='470m', ocodeu='470mΩ'),
    dict(maps=IEC60062_MAPS, icode='r47', q='470 mΩ', ocode='470m', ocodeu='470mΩ'),
    dict(maps=IEC60062_MAPS, icode='4r7', q='4.7 Ω', ocode='4r7', ocodeu='4r7Ω'),
    dict(maps=IEC60062_MAPS, icode='47r', q='47 Ω', ocode='47', ocodeu='47Ω'),
    dict(maps=IEC60062_MAPS, icode='470r', q='470 Ω', ocode='470', ocodeu='470Ω'),
    dict(maps=IEC60062_MAPS, icode='k47', q='470 Ω', ocode='470', ocodeu='470Ω'),
    dict(maps=IEC60062_MAPS, icode='K47', q='470 Ω', ocode='470', ocodeu='470Ω'),
    dict(maps=IEC60062_MAPS, icode='4k7', q='4.7 kΩ', ocode='4K7', ocodeu='4K7Ω'),
    dict(maps=IEC60062_MAPS, icode='4K7', q='4.7 kΩ', ocode='4K7', ocodeu='4K7Ω'),
    dict(maps=IEC60062_MAPS, icode='47k', q='47 kΩ', ocode='47K', ocodeu='47KΩ'),
    dict(maps=IEC60062_MAPS, icode='47K', q='47 kΩ', ocode='47K', ocodeu='47KΩ'),
    dict(maps=IEC60062_MAPS, icode='470k', q='470 kΩ', ocode='470K', ocodeu='470KΩ'),
    dict(maps=IEC60062_MAPS, icode='470K', q='470 kΩ', ocode='470K', ocodeu='470KΩ'),
    dict(maps=IEC60062_MAPS, icode='M470', q='470 kΩ', ocode='470K', ocodeu='470KΩ'),
    dict(maps=IEC60062_MAPS, icode='4M70', q='4.7 MΩ', ocode='4M7', ocodeu='4M7Ω'),
    dict(maps=IEC60062_MAPS, icode='47M', q='47 MΩ', ocode='47M', ocodeu='47MΩ'),
    dict(maps=IEC60062_MAPS, icode='470M', q='470 MΩ', ocode='470M', ocodeu='470MΩ'),
    dict(maps=IEC60062_MAPS, icode='G470', q='470 MΩ', ocode='470M', ocodeu='470MΩ'),
    dict(maps=IEC60062_MAPS, icode='4G7', q='4.7 GΩ', ocode='4G7', ocodeu='4G7Ω'),
    dict(maps=IEC60062_MAPS, icode='47G', q='47 GΩ', ocode='47G', ocodeu='47GΩ'),
    dict(maps=IEC60062_MAPS, icode='470G', q='470 GΩ', ocode='470G', ocodeu='470GΩ'),
    dict(maps=IEC60062_MAPS, icode='4700G', q='4.7 TΩ', ocode='4T7', ocodeu='4T7Ω'),  # uses invalid sf
]

def test_rkm_codes():
    # set maps to their default behavior
    set_prefs(rkm_maps=None, units_to_rkm_base_code=None, map_sf=None)

    # run tests
    for case in testcases:
        set_prefs(case['maps'])
        maps = 'unitless' if case['maps'] == UNITLESS_MAPS else 'IEC60062'
        q = from_rkm(case['icode'])
        print(
            maps,
            case['icode'],
            q,
            to_rkm(q),
            to_rkm(q, show_units=True),
            sep=', '
        )
        assert str(q) == case['q']
        assert to_rkm(q) == case['ocode']
        assert to_rkm(q, show_units=True) == case['ocodeu']

    # return maps to their default behavior
    set_prefs(rkm_maps=None, units_to_rkm_base_code=None, map_sf=None)


def test_misc():
    # set maps to their default behavior
    set_prefs(rkm_maps=None, units_to_rkm_base_code=None, map_sf=None)

    c = Quantity(2)
    assert to_rkm(c) == '2'
    assert to_rkm(c, strip_zeros=False) == '2d0'
    assert to_rkm(c, strip_zeros=False, prec=5) == '2d00000'

    c = Quantity(20)
    assert to_rkm(c) == '20'
    assert to_rkm(c, strip_zeros=False) == '20'
    assert to_rkm(c, strip_zeros=False, prec=5) == '20d0000'

    c = Quantity(2, units='Ω')
    assert to_rkm(c) == '2'
    assert to_rkm(c, strip_zeros=False) == '2r0'
    assert to_rkm(c, strip_zeros=False, prec=5) == '2r00000'

    c = Quantity(20, units='Ω')
    assert to_rkm(c) == '20'
    assert to_rkm(c, strip_zeros=False) == '20'
    assert to_rkm(c, strip_zeros=False, prec=5) == '20r0000'


rkm_codes/rkm_codes.py
# RKM codes
# encoding: utf8

# description {{{1
"""
RKM codes are used to represent electrical quantities in labels, particularly on
schematics and on the components themselves.  They are standardized in various
national and international standards, including: IEC 60062 (1952) (formerly IEC 62),
DIN 40825 (1973), BS 1852 (1974), IS 8186 (1976) and EN 60062 (1993).
IEC-60062 was significantly updated in 2016.

RKM codes were originally meant also as part marking code.  This shorthand
notation is widely used in electrical engineering to denote the values of
resistors and capacitors in circuit diagrams and in the production of electronic
circuits (for example in bills of material and in silk screens). This method
avoids overlooking the decimal separator, which may not be rendered reliably on
components or when duplicating documents.  They also provide the benefit that
the characters within a RKM code are either letters or digits, and so can be
embedded within identifiers without introducing invalid characters.

IEC 60062 is described in https://en.wikipedia.org/wiki/RKM_code.

Essentially an RKM version of a number is the number with a scale factor where
the decimal point replaced by the scale factor. For example, a resistance of
4.7kΩ becomes 4k7. If there is no scale factor, the decimal point is replaced by
a letter that signifies the type of the component.  For example, a resistance of
4.7Ω becomes 4r7.

In the standard, large values are assumed to be resistances and small values are
assumed to be capacitances.  So 4k7 is a resistance and 2n5 is a capacitance.
However, this package also supports a version of RKM codes where the units are
not implied by the value, making RKM codes suitable for a wider variety of value
types, such as voltage, current, and inductance.
"""

# license {{{1
# Copyright (C) 2018 Kenneth S. Kundert
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see [http://www.gnu.org/licenses/].


# imports {{{1
import re
from quantiphy import Quantity

# constants {{{1
# version {{{2
__version__ = '0.4.0'
__released__ = '2019-08-29'

# IEC60062 maps {{{2
IEC60062_MAPS = {
    # These maps conform to the latest recommended mappings as defined in the 
    # IEC-60062 standard. They apply assume only resistors and capacitors. Small 
    # values are associated with capacitors and large values with resistors. The 
    # units can be inferred by the scale factors and so it is unnecessary to 
    # specify the units with the number. For example, 4k7 is a resistance (k 
    # implies resistance) and 2p5 is a capacitance (p implies capacitance)
    #
    # These maps map a RKM code into a scale factor and the implied units (the 
    # implied units can be overwritten by explicitly specified units)

    # Resistors
    'r': ('', 'Ω'),
    'R': ('', 'Ω'),
    'Ω': ('', 'Ω'),
    'L': ('m', 'Ω'),
    'l': ('m', 'Ω'),
    'k': ('k', 'Ω'),
    'K': ('k', 'Ω'),
    'M': ('M', 'Ω'),
    'G': ('G', 'Ω'),
    'T': ('T', 'Ω'),

    # Capacitors
    'c': ('', 'F'),
    'C': ('', 'F'),
    'F': ('', 'F'),
    'm': ('m', 'F'),
    'm': ('m', 'F'),
    'u': ('μ', 'F'),
    'μ': ('μ', 'F'),
    'n': ('n', 'F'),
    'p': ('p', 'F'),
}

# unitless maps {{{2
UNITLESS_MAPS = {
    # These maps do not conform to the IEC-60062 standard, but can be applied to 
    # values other than resistance and capacitance. In this case, adding 
    # optional units is helpful because, while it would be possible to infer the 
    # units for some values (such as 4r7), it is not possible for most values 
    # (such as 4m7) and so implied units are not specified for any value.
    #
    # These maps map a RKM code into a scale factor and the implied units (the 
    # implied units can be overwritten by explicitly specified units)

    'r': ('', ''),  # resistors
    'R': ('', ''),
    'v': ('', 'V'),  # voltages
    'i': ('', ''),  # currents
    'I': ('', ''),
    'c': ('', ''),  # capacitors
    'f': ('', ''),
    'l': ('', ''),  # inductors
    'h': ('', ''),
}

# map units to rkm base code {{{2
# Controls the base code produced by to_rkm().
UNITS_TO_RKM_BASE_CODE = {
    'Ω': 'r',
    'Ohm': 'r',
    'F': 'c',
    'H': 'l',
    'V': 'v',
    'A': 'i',
}

# map scale factors {{{2
# Controls the scale factors produced by to_rkm().
MAP_SF = dict(u='μ', k='K')
MINUS_SIGN = 'n'
SHOW_UNITS = False
STRIP_ZEROS = True
PREC = 1

# utilities {{{1
# cull {{{2
def cull(collection):
    return (v for v in collection if v)

# set_prefs {{{1
_rkm_maps = UNITLESS_MAPS
_units_to_rkm_base_code = UNITS_TO_RKM_BASE_CODE
_map_sf = MAP_SF
_minus_sign = MINUS_SIGN
_show_units = SHOW_UNITS
_prec = PREC
_strip_zeros = STRIP_ZEROS

def set_prefs(
    rkm_maps=False, units_to_rkm_base_code=False, map_sf=False,
    show_units=0, strip_zeros=0, minus_sign=False, prec=False
):
    '''Set Preferences

    Use to set values that control the behavior of the RKM code.
    Any values not passed in a left alone.
    Pass in *None* to reset a preference to its default value.

    Args:
        rkm_maps (dictionary of tuples):
            A dictionary that maps a base code or scale factor into a scale
            factor and units. Used to affect the behavior or *from_rkm()*.
            Generally set to rkm_codes.IEC60062_MAPS, which encourages
            conformance to the standard, or to rkm_codes.UNITLESSs_MAPS, which
            supports a wider range of quantities than resistances and
            capacitances. You can also pass in a custom mapping to get a
            particular result.

        units_to_rkm_base_code (dictionary of strings):
            A dictionary that maps a units to a base code. Used to affect the
            behavior of *to_rkm()*.

        map_sf (dictionary of strings):
            A dictionary that maps the scale factors used by QuantiPhy to the
            ones found in a RKM code. Used to affect the
            behavior of *to_rkm()*.
        show_units (bool):
            Whether the units should be included in the RKM code.
        stripzeros (bool):
            Whether the units should be included in the RKM code.
        minus_sign (str):
            Character to use to indicate a negative value. The default is 'n',
            but 'm' is also common.
        prec (int):
            Precision. Number of digits shown in 1 plus this number.
    '''
    global _rkm_maps, _units_to_rkm_base_code, _map_sf
    global _show_units, _strip_zeros, _minus_sign, _prec

    if rkm_maps is None:
        _rkm_maps = UNITLESS_MAPS
    elif rkm_maps is not False:
        _rkm_maps = rkm_maps

    if units_to_rkm_base_code is None:
        _units_to_rkm_base_code = UNITS_TO_RKM_BASE_CODE
    elif units_to_rkm_base_code is not False:
        _units_to_rkm_base_code = units_to_rkm_base_code

    if map_sf is None:
        _map_sf = MAP_SF
    elif map_sf is not False:
        _map_sf = map_sf

    if show_units is None:
        _show_units = SHOW_UNITS
    elif show_units is not 0:
        _show_units = show_units

    if strip_zeros is None:
        _strip_zeros = STRIP_ZEROS
    elif strip_zeros is not 0:
        _strip_zeros = strip_zeros

    if minus_sign is None:
        _minus_sign = MINUS_SIGN
    elif minus_sign is not False:
        _minus_sign = minus_sign

    if prec is None:
        _prec = PREC
    elif prec is not False:
        _prec = prec

# RKM code patterns {{{1
# regex1 matches rkm codes that start with a digit.
# regex2 matches rkm codes that end with a digit.
ld_regex = r'([pmn]?)([0-9]+)([a-zµμΩ℧]+)([0-9]*)'
td_regex = r'([pmn]?)([0-9]*)([a-zµμΩ℧]+)([0-9]+)'
combined_regex = '(?:{})|(?:{})'.format(ld_regex, td_regex)
ld_matcher = re.compile(ld_regex, re.I)
td_matcher = re.compile(td_regex, re.I)
combined_matcher = re.compile(combined_regex, re.I)

# from_rkm {{{1
def from_rkm(code):
    '''From RKM

    Convert a RKM code string to a quantiphy.Quantity.

    Args:
        code (str):
            An RKM code that may include explicitly specified. Examples of
            acceptable RKM codes for resistance include:   R47 (0.47 Ω), 4R7
            (4.7 Ω), 470R (470 Ω), 4K7 (4.7 kΩ), 47K (47 kΩ), 47K3 (47.3 kΩ),
            470K (470 kΩ), and 4M7 (4.7 MΩ).  Units may be added by appending
            the units to the scale factor, as in 47KΩ3 (47.3 kΩ).
    Returns:
        A quantiphy.Quantity if a valid RKM code was found, otherwise *None* is
        returned.
    '''
    for matcher in [ld_matcher, td_matcher]:
        match = matcher.match(code)
        if match:
            sign, ld, base, td = match.groups()
            sf, units = _rkm_maps.get(base, (base, ''))
            if sign and sign in 'mn':
                sign = '-'
            else:
                sign = ''
            return Quantity(''.join(cull([sign, ld,  '.',  td,  sf, units])))

# to_rkm {{{1
def to_rkm(q, prec=None, show_units=None, strip_zeros=None):
    '''To RKM

    Convert a quantiphy.Quantity to an RKM string.

    Args:
        q (quantiphy.Quantity):
            The value to be converted to an RKM code.
        prec (int):
            The precision. The number of digits is the precision + 1.
        show_units (bool):
            Whether and where the units should be included in the RKM code
            (default is False).
        strip_zeros (bool):
            Whether excess zeros should be removed (default is True).
    Returns:
        A quantiphy.Quantity if a valid RKM code was found, otherwise *None* is
        returned.
    '''
    if show_units is None:
        show_units = _show_units
    if strip_zeros is None:
        strip_zeros = _strip_zeros
    if prec is None:
        prec = _prec
    units = q.units if show_units else ''
    rkm_base_code = _units_to_rkm_base_code.get(q.units, q.units)

    value = q.render(
        form='si', show_units=False, strip_zeros=False, strip_radix=False, 
        prec=prec
    )
    value = value.replace('-', _minus_sign)
    sf = value[-1]
    if sf and sf in q.output_sf:
        value = value[:-1]
    elif units:
        sf = ''
    else:
        sf = rkm_base_code
    if '.' not in value:
        value += '.'
    if strip_zeros:
        value = value.rstrip('0')
    if not sf and value[-1] != '.':
        if units:
            sf = units
            units = ''
        else:
            sf = 'd'
    if sf == rkm_base_code:
        value = value.rstrip('.')
    return value.replace('.', _map_sf.get(sf, sf)+units)

# find_rkm {{{1
def find_rkm(text, kind=None):
    '''Find RKM codes

    Iterate through RKM codes found in a text string.

    Args:
        text (str):
            The text string that is to be searched for RKM codes.
        kind (str):
            Specify:
                'ld': match only RKMs with leading digits
                'td': match only RKMs with trailing digits
                anything else: match either type of RKMs
    Yields:
        Succession of quantities.
    '''
    if kind == 'ld':
        matcher = ld_matcher
    elif kind == 'td':
        matcher = td_matcher
    else:
        matcher = combined_matcher

    for match in matcher.finditer(text):
        yield from_rkm(match.expand(r'\g<0>'))
